### DISCUSSION TITLE
DrawText with multiregion unicode chars

### DISCUSSION THREAD

**Discussion by TomQv** [mono/SkiaSharp#2761] [Q&A]
As discussed here #232 and here #435 Skia can't render text with multiple unicode regions.
At least not 6 years ago, but it seems, this didn't change until now.  

For example "Oum Dreyga ⵓⵎ ⴷⵔⵉⴳⴰ أم ادريكة " which is a place in Morocco called Oum Dreyga. 
Additionally the name is printed also in local and arabic language.

To print this with SKCanvas.DrawText we would need a Font which has all characters includes, which I didn't find so far.
I only found this http://unifoundry.com/unifont/index.html, but this is a bitmap-font, not so nice.

Second option would be, if DrawText could just ignore characters, which are not in the current typeface, so that these ugly blocks were not rendered: 
![grafik](https://github.com/mono/SkiaSharp/assets/6605960/443c0ad8-8442-481d-a357-75f9fd8ad37a)
![grafik](https://github.com/mono/SkiaSharp/assets/6605960/1b58bb02-6352-444c-9f14-e3b120e174d2)

Is there maybe a setting to ignore unknown chars?

**Comment by Gillibald** [comment]
https://github.com/toptensoftware/RichTextKit

**Reply by taublast** [reply]
https://github.com/taublast/DrawnUi.Maui.Demo/blob/main/src/Usual/MainPage.xaml

`SkiaMarkDownLabel` will auto-split your text into spans with the corresponding fonts to render different unicode chunks, you can use it in a usual MAUI app or grab the source code to integrate in your skiasharp project.

**Reply by Gillibald** [reply]
Your solution just uses a naive 1:1 character to glyph mapping and does no extra processing. No shaping to bidi analysis. So in the end you just support the Latin writing system. [RichTextKit](https://github.com/toptensoftware/RichTextKit) supports all of this.

**Reply by taublast** [reply]
Thanks for taking a look :) Actually besides drawings glyphs it detects if the chunk needs shaping and draws it as shaped text in that case, what is required for Arabic etc.

![image](https://github.com/mono/SkiaSharp/assets/25801194/944c9b1d-3b12-4cca-b5ab-7b004c7356bc)

**Reply by taublast** [reply]
There is a nuget for xamarin too, no uwp, android+ios only: https://github.com/taublast/DrawnUi.Xamarin

**Reply by Gillibald** [reply]
Wasn't clear you are mapping Unicode regions to a code path that shapes the text. Ideally, everything is shaped because you are missing font features that someone might want.

**Comment by TomQv** [comment]
Thanks, great stuff. 
Quick test with Xamarin.Forms works good.
But unfortunate only on UWP.
Android seems to have problems with devicescale. Didn't test iOS yet.

I would prefer a setting in SkiaSharp just to ignore unprintable stuff.
